app/models/article.rb

```diff
@@ -99,6 +98,9 @@ class Article < ApplicationRecord
   validate :validate_co_authors, unless: -> { co_author_ids.blank? }
   validate :validate_co_authors_must_not_be_the_same, unless: -> { co_author_ids.blank? }
   validate :validate_co_authors_exist, unless: -> { co_author_ids.blank? }
+  # these two are order dependent since validate_title changes the input
```

Comment:
Seems like `validate_title` should actually be a `before_validation` callback, it doesn't seem to actually validate anything (as in adding to `errors`). That would definitely feel more idiomatic than order-dependent validations.

Something like this:

```ruby
before_validation :remove_prohibited_unicode_characters

def remove_prohibited_unicode_characters
  return unless title.match?(PROHIBITED_UNICODE_CHARACTERS_REGEX)

  self.title = title.gsub(PROHIBITED_UNICODE_CHARACTERS_REGEX, "")
end
```

This would also get rid of one extra method (`contains_prohibited_unicode_characters?`). Admittedly this loses the option of replacing the characters with anything other than the empty string but we're not currently using this anyway so YAGNI?

Reply:
thanks - the "before validation" callback approach does make sense (and was part of the problem I foresaw originally).